### ISSUE TITLE
Incorrectly calculated order total after applying discount coupon at Edit Order screen

### ISSUE THREAD

**Issue by starypop** [woocommerce/woocommerce#25447]
**Describe the bug**
When manually applying a coupon at Edit Order admin screen to an order that has a product with a specific price and tax rate (ie. $129.00 and 8% tax rate), the recalculated order total is 0.01 lower than it should be. Having the same Store settings, the coupon discount and discounted order totals calculations in front-end Cart and Checkout pages are done correctly.

**To Reproduce**
1. Check "Yes, I will enter prices inclusive of tax" option in Taxes settings.
2. Set "Including tax" option for "Display prices in the shop" and "Display prices during cart and checkout" selects.
3. Add a Standard tax rate of 8%.
4. Create a coupon "test" having "Fixed cart discount" type and amount of 10.
5. Create a simple taxable product "Test Product" with regular price 129 and set its Tax class to "Standard".
6. Add "Test Product" to cart.
7. In cart, apply the "test" coupon.
8. Go through the whole front-end checkout process and see that the order discount amount as well as the order total are calculated correctly everywhere (in Cart, Checkout, Thank-you Page and in admin's orders list it's: 129.00-10=119.00).
9. Try to replicate the same order by creating a new order in WP Admin.
10. Add "Test Product" to this order as an order item and recalculate totals.
11. Apply coupon "test" and recalculate totals.
12. Notice that the recalculated order total is now incorrect by 0.01 (the order total here is 118.99 instead of 119.00).

**Screenshots**
Correct totals in Checkout:
![Zrzut ekranu 2020-01-21 o 15 33 54](https://user-images.githubusercontent.com/27000783/72817357-e934b680-3c69-11ea-8566-2fd3a55f440f.png)

Incorrect totals in WP Admin after applying a coupon using this screen:
![Zrzut ekranu 2020-01-21 o 15 36 15](https://user-images.githubusercontent.com/27000783/72817422-07021b80-3c6a-11ea-8cc3-790efa1a1efa.png)

**Expected behavior**
All orders that have the same coupon should have the same, consistent order total. In the example shown above, when applying a coupon in WP Admin, order total is 0.01 lower than it should be.

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
<details>
`
### WordPress Environment ###

WordPress address (URL): http://localhost:8888/wc
Site address (URL): http://localhost:8888/wc
WC Version: 3.8.1
REST API Version: ✔ 1.0.2
Log Directory Writable: ✔
WP Version: 5.3.2
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.2.34 (Unix) mod_wsgi/3.5 Python/2.7.13 PHP/7.2.1 mod_ssl/2.2.34 OpenSSL/1.0.2j DAV/2 mod_fastcgi/2.4.6 mod_perl/2.0.9 Perl/v5.24.0
PHP Version: 7.2.1
PHP Post Max Size: 8 MB
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 7.52.1
OpenSSL/1.0.2j

SUHOSIN Installed: –
MySQL Version: 5.6.38
Max Upload Size: 8 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.8.1
WC Database Prefix: wp_
MaxMind GeoIP Database: ❌ The MaxMind GeoIP Database does not exist - Geolocation will not function. You can download and install it manually from https://dev.maxmind.com/geoip/geoip2/geolite2/ to the path: . Scroll down to "Downloads" and download the "MaxMind DB binary
gzipped" file next to "GeoLite2 Country". Please remember to uncompress GeoLite2-Country_xxxxxxxx.tar.gz and upload the GeoLite2-Country.mmdb file only.

Total Database Size: 3.06MB
Database Data Size: 2.08MB
Database Index Size: 0.98MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 1.50MB + Index: 0.03MB + Engine InnoDB
wp_postmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_posts: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (1) ###

WooCommerce: by Automattic – 3.8.1

### Settings ###

API Enabled: –
Force SSL: –
Currency: GBP (£)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –

### WC Pages ###

Shop base: #6 - /shop/
Cart: #7 - /cart/
Checkout: #8 - /checkout/
My account: #9 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty
Version: 1.1
Author URL: https://wordpress.org/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Action Scheduler ###

Complete: 0
Oldest: –
Newest: –

Pending: 0
Oldest: –
Newest: –

Canceled: 0
Oldest: –
Newest: –

In-progress: 0
Oldest: –
Newest: –

Failed: 0
Oldest: –
Newest: –

`
</details>

**Comment by starypop** [comment]
In the meantime a new 3.9 WC version has been released.

Unfortunately, after testing it I can confirm that the bug **is still there**.

**Comment by tammullen** [comment]
Hi @starypop,

Thank you for taking the time reporting this bug, we really appreciate your help. I can reproduce it on my end by following the steps you listed.

![Screenshot 2020-01-27 at 11 47 10](https://user-images.githubusercontent.com/24649833/73172947-1b2b9a00-40fc-11ea-85da-77e7be23d6a1.png)
![Screenshot 2020-01-27 at 11 48 25](https://user-images.githubusercontent.com/24649833/73172944-1b2b9a00-40fc-11ea-996e-9854e15b1f99.png)

However I am unable to so far recreate it using other products or using the same simple test product but set to a different price (both lower and higher than the 129.00 amount in your example). 

![Screenshot 2020-01-27 at 12 06 04](https://user-images.githubusercontent.com/24649833/73173724-f20c0900-40fd-11ea-8c62-f4aff5afce11.png)
![Screenshot 2020-01-27 at 12 05 32](https://user-images.githubusercontent.com/24649833/73173725-f20c0900-40fd-11ea-806e-9d7089bc2895.png)


As it doesn't appear to be affecting all orders created in the back end I am setting the priority to Low for the moment. 
This means we won’t be able to include this fix in the upcoming release due to the lower priority of this issue compared to others reported. We’re going to add it to our backlog so we can include it in our planning for one of our future releases.

If you have any other examples that can be reproduced let us know.

**Comment by starypop** [comment]
Thanks @tammullen for your answer!

I understand that applying a coupon via admin Edit Order screen might not be that important feature at the moment.

Unfortunately the bug seems to be somewhere deeper than the admin Edit Order interface. I've looked the method WC_AJAX::add_order_item() that is hooked to the 'woocommerce_add_order_item' AJAX action which was fired when inserting products to an order by the modal. I'm afraid that **the bug also occurs everywhere where WC_Order methods are being used** (and not only by WooCommerce core, it is probably affecting plugins as well).

> If you have any other examples that can be reproduced let us know.

I've made a simple testing script, that checks the number of this bug occurrences in bulk. Having the same 'Test Product' and 'test' coupon that I've described in the original post, it iterates for a given number and sets the product's price plus 1. Please have in mind that it checks the 8% tax and prices that has no decimals only. The code of the script is using the same WC_Order methods to apply items, coupons and recalculating totals as WC_AJAX::add_order_item() does. 

While testing it from 1 to 1000 it gave me the results of 111/1000 bugged lines. **That is over 10%** of all possible combinations, which seems to be not so low number to be honest… 

```
$max_price = 1000;

// Create an order
$order = new WC_Order();
$order->set_prices_include_tax( 'yes' === get_option( 'woocommerce_prices_include_tax' ) );
$order->save();

// The simple Test Product with tax 8%
$product_id = 10;
$product = wc_get_product( $product_id );

// The fixed cart coupon with amount of 10
$coupon_code = 'test';
$coupon = new WC_Coupon( $coupon_code );
$coupon_amount = $coupon->get_amount();

$bug_needle = '.00';
$bug_needle_lenght = strlen( $bug_needle );
$bugged_lines = array();

for ( $i=1; $i <= $max_price; $i++ ) :

	// Clear order
	$order->remove_order_items();
	$order->calculate_totals( true );

	// Set product's new price
	$product->set_price( $i );

	// Insert new item
	$item_id = $order->add_product( $product, 1 );

	// Apply coupon (it also recalculates order total once again)
	$order->apply_coupon( $coupon );

	// Check if there's a bug on this iteration (based on order_total decimals amount)
	if( strlen( $order->get_total() ) - $bug_needle_lenght != strrpos( $order->get_total(), $bug_needle ) && $i > $coupon_amount ) {
		$expected_total = wc_format_decimal( ( $i - $coupon_amount ), '' );
		$bugged_lines[] = $expected_total . ' -> ' . $order->get_total();
	}

endfor;

print_r( $bugged_lines );
echo 'Bugs total: ' . count( $bugged_lines ) . '/' . $max_price;
```

**Comment by starypop** [comment]
Hey @tammullen, @juliaamosova,

Could you take a look at my answer and detailed explanations of this bug that I've made and maybe review this bug's priority?

Or should I create another issue, as I've found the bug also exists in WC_Order class methods?

Thanks for your help!

**Comment by starypop** [comment]
Hey @vedanshujain, could you please take a look at this issue and verify if it has been fixed by your PR #25800? Thanks!

**Comment by vedanshujain** [comment]
@starypop Looks like this issue is indeed fixed by #25800 so I am closing it. Also many thanks for the code script, it gave me a few ideas on how we can make our unit tests more robust to detect off by one errors.
